content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1118 @@
+* [Quiz](./quiz/sections-always-shared.md)
+
+![Copy-on-write after Write](./media/copy-on-write-final.svg)
+
+----
+
+### Copy-on-write Overheard
+
+* Writes cause minor page faults
+
+```
+student@os:~/.../compute/lecture/demo/copy-on-write$ ./copy_on_write_overhead 
+ * Child process started
+ -- Press ENTER to continue ...
+Time for reading 131072 pages: 30 ms
+ * Child process read pages
+ -- Press ENTER to continue ...
+Time for writing to 131072 pages: 222 ms
+[...]
+```
+
+* The child process performs exactly 131072 **minor** page faults when writing data
+
+```
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # before writing
+22
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # after writing
+131094
+```
+
+---
+
+## Threads
+
+----
+
+## Threads - Backstory
+
+* You have an application that applies filters to a set of images
+* If we used processes to spread the work:
+    * images (like all data) would be marked copy-on-write
+    * processes cannot easily modify the "shared" images
+* **Solution:** use threads
+    * No copy-on-write: threads share the heap and data sections
+
+----
+
+### Threads - Definitions
+
+* Scheduling unit
+* Lightweight processes (LWP)
+* Each process has at least one thread
+
+---
+
+### Thread - Resources
+
+![Thread Resources](media/process-vs-thread.png)
+
+----
+
+### App, Process, Thread - The Bigger Picture
+
+* [Quiz](./quiz/threads-shared-data.md)
+
+![Application Architecture](media/app-process-thread.svg)
+
+----
+
+### `fork` and Threads
+
+* `fork` also creates a new thread...
+* ... but part of a different process
+
+![fork - new Thread](./media/fork-thread.svg)
+
+---
+
+### Process vs Thread
+
+| PROCESS                          | THREAD                               |
+| :------------------------------- | :----------------------------------- |
+| independent                      | part of a process                    |
+| isolated VAS                     | shares VAS with other threads        |
+| slower creation                  | faster creation                      |
+| larger context switch duration   | shorter context switch duration      |
+| ending means ending all threads  | other threads continue when finished |
+
+----
+
+### Process vs Thread
+
+`demo/create-thread/create_thread.py`
+
+vs
+
+`demo/create-process/create_process.py`
+
+---
+
+## Scheduling
+
+----
+
+### Scheduling - Why?
+
+* For multitasking: more threads than CPU cores
+* For responsiveness: small waiting time
+* For productivity: execute more threads
+
+----
+
+### Typical Thread Execution 
+
+![Thread Execution](media/thread-behaviour.png)
+
+----
+
+### Thread States
+
+* At most one RUNNING thread per core
+* Queues for BLOCKED and READY threads
+* [Quiz](./quiz/number-of-running-threads.md)
+
+![Thread State Diagram](media/thread-states.svg)
```

Comment:
Can you make this image smaller?

Reply:
Done, but had to use inline HTML. I think it's alright since it also renders fine in Markdown.